user-auth/enclave/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+            .unwrap();
+        user_id_buf
+            .set(slice::from_raw_parts(user_id, user_id_length))
+            .unwrap();
+        salt_buf
+            .set(std::str::from_utf8(slice::from_raw_parts(salt, SALT_LENGTH)).unwrap())
+            .unwrap();
+    }
+
+    let pepper = new_pepper(user_id_buf.get().unwrap());
+    let hash_ctx = new_hash_context(pepper.as_ref());
+
+    match hash_ctx {
+        Err(_) => sgx_status_t::SGX_ERROR_UNEXPECTED,
+        Ok(ctx) => ctx
+            .hash_password(password_buf.get().unwrap(), *salt_buf.get().unwrap())
+            .map_or_else(
+                |_| sgx_status_t::SGX_ERROR_UNEXPECTED,
+                |hash| {
+                    unsafe {
+                        hash_string.copy_from_nonoverlapping(
+                            hash.serialize().as_str().as_bytes().as_ptr(),
+                            HASH_STRING_LENGTH,
+                        );
+                    }
+                    sgx_status_t::SGX_SUCCESS
+                },
+            ),
+    }
+}
+
+///# Safety: guaranteed by the trusted bridge routine generated with the
+/// Intel SDK's `sgx_edger8r` tool
+#[no_mangle]
+pub unsafe extern "C" fn verify_password(
+    // input paramaters
+    password: *const u8,
+    password_length: usize,
+    user_id: *const u8,
+    user_id_length: usize,
+    stored_hash_string: *const u8,
+    // output paramaters
+    verify_status: *mut VerifyPasswordStatus,
+) -> sgx_status_t {
+    let password_buf = OnceCell::<&[u8]>::new();
+    let user_id_buf = OnceCell::<&[u8]>::new();
+    let hash_string_buf = OnceCell::<&[u8]>::new();
+    unsafe {
+        password_buf
+            .set(slice::from_raw_parts(password, password_length))
+            .unwrap();
+        user_id_buf
+            .set(slice::from_raw_parts(user_id, user_id_length))
+            .unwrap();
+        hash_string_buf
+            .set(slice::from_raw_parts(
+                stored_hash_string,
+                HASH_STRING_LENGTH,
+            ))
+            .unwrap();
+    }
+
+    let hash_string = std::str::from_utf8(hash_string_buf.get().unwrap()).unwrap();
+    let pepper = new_pepper(user_id_buf.get().unwrap());
+    new_hash_context(pepper.as_ref()).map_or_else(
+        |_| sgx_status_t::SGX_ERROR_UNEXPECTED,
+        |ctx| match ctx.verify_password(
+            password_buf.get().unwrap(),
+            &PasswordHashString::parse(hash_string, password_hash::Encoding::default())
+                .unwrap()
+                .password_hash(),
+        ) {
+            Ok(()) => {
+                unsafe { *verify_status = VerifyPasswordStatus::PasswordVerified };
+                sgx_status_t::SGX_SUCCESS
+            }
+            Err(password_hash::Error::Password) => {
+                unsafe { *verify_status = VerifyPasswordStatus::InvalidPassword };
+                sgx_status_t::SGX_SUCCESS
+            }
+            Err(_) => sgx_status_t::SGX_ERROR_UNEXPECTED,
```

Comment:
I would at least do `eprintln!` ocall, or debugging this may get unpleasant, requiring a re-compile. We can also do a TODO note, so we don't forget to look at it again when it gets to production.